zepez/zms
# ✨ Video transocding and hls.js

This pr introduces the first version of transcoding logic with video playback.

We are using ffmpeg to do the transcoding to HLS format. HLS allows us to stream the video over http in segments, which will provide better playback performance than vanilla html video solutions. 

To test this pr: 

- Make sure ffmpeg is installed on your machine. On MacOS, this can be done with homebrew: `brew install ffmpeg`
- Add your video file to `media/movies/example.mp4`
- Navigate to `packages/transcode/src/index.ts` and reference your video file `media/movies/example.mp4` in `pipeline()`
- Run the script `pnpm esno packages/tracnsode/src/index.ts`
- Start the web server `pnpm dev`
- Navigate to `http://localhost:3000/movies/play/example.mp4`

I've made the transcoding logic asynchronous, which means that we will create the HLS playlists in this order: 360p > 720p > 1080p. 

I've made each playlist available at the start of the transcoding process. This allows for immediate video playback at 360p and means we're effectively "live transcoding" the video (only applicable if playback is requested immediately before transcoding is finished). 

Once transcoding is finished, the highest quality is automatically played based on the client's available bandwidth. 

I have also started the basic custom UI for the video player. This includes: 
- progress bar with buffer indicator
- volume / mute control
- timecodes 
- play/pause control and indicators 
- fullscreen control
